### PR TITLE
Probabilistic continuous feature matching

### LINKED ISSUE #1119: Better matching for continuous features
We want to implement better matching for continuous features, mainly to support better action learning. We need to be able to match distances to learn constraints for certain actions, e.g. "to take something you have to be close to it". The current approach isn't ideal -- we'd like to _learn_ a range of values for how close we have to be, rather than fixing a tolerance.

I'm including a work-in-progress plan. I'm going to edit this post as that plan changes to keep things in one place for our own sanity.

# [Rough design](https://github.com/isi-vista/adam/issues/1102#issuecomment-1093361068)
> So during the learning stage every introduction of a new distance would add to a running mean, standard deviation, max, and min values. (max and min is mostly for being able to display those values... I'm unsure they are strictly needed for determining a match). Come time for matching a match is made if the determined values has a greater than X% chance of being in the distribution formed by the mean and std deviation. This % chance should probably be set by a parameter for the moment. Ideally this value would instead contribute to the 'match score' in some way but that's a larger technical challenge than I want to take on if we don't have to.

Some questions this raises:

1. How do we calculate a match percentage? Mean and variance alone don't give us any sort of percent match. Two problems here:
    1. We don't have a distributional form and no general one may exist. This isn't fatal, though -- there are a few different methods we could reasonably try to estimate a distribution. Like: Use a Gaussian, use percentiles, or use kernel density estimation (KDE) maybe with a Gaussian kernel. And I'm sure there are other methods we could try.
    2. Even if we assume a particular distributional form, we don't really have a match percentage. The closest thing we have is something like a hypothesis test p-value, "how consistent is the given value with this distribution?" But we do have that value, and I think that seems like a reasonable first attempt.
3. How do we perform updates? I think we want to do this in an online/constant time way that does not weight old examples too heavily (but also doesn't overweight new ones). This should probably factor into how we calculate match percentages.
4. How do we parameterize the % chance threshold? In the current scheme I think this has to go in the node because otherwise the node wouldn't have access to the threshold at match time. We don't yet have a better way to parameterize things, I think.
5. How do we calculate match percentages for and update on pattern-pattern matches? In general this seems to require calculating a similarity between two distributions. For now we _could_ ignore the problem and just throw an error whenever we try to match two same-label continuous features that each have more than one observation. However, this may cause problems for contrastive learning (see section below on the "tie-in problems").

# Issues related to Gaussian match percentages
Gaussian match percentages seem simple but pose some problems, particularly with updating. We have options:

1. Running average/variance over all samples. We can calculate these efficiently using [Welford's algorithm](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm). However, this weights all examples equally, which could be a problem if for example we were to train ADAM on a bunch of examples and then later need to adapt ADAM to some arbitrarily chosen new examples.
2. Exponential smoothing, i.e. using an exponential moving average. This works fine for estimating the mean, but for the variance I haven't yet found an equivalent.

Re: estimating variance with exponential smoothing, I did some searching but didn't find anything obvious. The closest relatives I could think of were BatchNorm and the Adam optimizer. However, neither seems like a good fit. The BatchNorm paper doesn't describe any such thing; PyTorch seems to track a running mean and variance but I haven't yet found the details on how it calculates those. Meanwhile, the Adam optimizer calculates a running uncentered second moment rather than a variance/centered second moment, and I don't know of a way to calculate the centered moment from the uncentered one, so that doesn't seem helpful.

I plan to look a bit more into PyTorch's BatchNorm and also see if I can quickly find a way of getting centered variance from uncentered variance.

# Tie-in problems with #1102
Some problems here tie in with #1102.

First, there's pattern-pattern matching. To do contrastive learning using graph matching, I think we have to either (1) intersect each perception graph with the other perception graph, (2) intersect each pattern with the other pattern. Alternatively we could (3) match each pattern to both perception graphs. Currently we do (1) which I worry will have wonky results with or without the new matching system because we can't properly estimate a distribution to match from single graphs (variance is the big issue). Doing (2) would require handling continuous pattern-pattern matching properly. Doing (3) might work, but is different enough to give me a headache so I would have to think about whether this approach makes sense.

Second, there's the issue of how to update continuous nodes in intersections. I don't think we want this to mutate the input pattern; that seems to go against the contract for intersection. We probably want to handle these similarly to a "regular" match, except it confirms the matches on a copy of the original and returns the updated pattern. That complicates #1102. However, given we've already bitten that bullet in other ways this is probably okay.

# Plan
1. Add to `NodePredicate` an abstract method `confirm_match(matched_with: Union[NodePredicate, PerceptionGraphNode]) -> None`. For all existing predicates this does nothing.
5. Define a new protocol `ContinuousValueMatcher`. This provides the following methods:
    1. `match_score(value: float) -> float` -- should be between 0 and 1 in current thinking, though I think this is mainly for conceptual convenience
    2. `similarity(matcher: ContinuousValueMatcher) -> float` -- this ranges from 0 to 1 -- I suspect this will be problematic to compute in general, so probably we'll just limit ourselves to same-type similarity and raise an error if you ask for something more complicated. This should be symmetric in the sense that `a.similarity(b) == b.similarity(a)` no matter what `a` and `b` you use -- maybe with some allowance for floating point errors.
    6. `merged_with(matcher: ContinuousValueMatcher) -> ContinuousValueMatcher` -- again, I think we can limit ourselves to same-type merging
    7. `update_on_observation(value: float) -> None`
3. Define a concrete implementation of `ContinuousValueMatcher` that we can use, which implements assume-it's-Gaussian matching. Use [Welford's algorithm](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm) to update on single observations. Similarity is left unimplemented and raises an error as we may not even need it given we are ignoring the distributions when matching distribution nodes. Merging is implemented only when one of the matchers has just a single observation, and otherwise raises `NotImplementedError()`. There should be a way to merge them but I'd have to work out the math for handling the variance merge and we don't need it yet so leaving it undefined for now seems OK. Some links that might be relevant for this later: [math exchange answer 1](https://math.stackexchange.com/a/2971522), [answer 2](https://math.stackexchange.com/a/1952759), [sketchy page with similar formula](https://www.emathzone.com/tutorials/basic-statistics/combined-variance.html).
4. Define a new class `DistributionalContinuousNodePredicate` which has a `label`, `matcher`, and `min_match_score`. This node is **mutable**.
    1. `__call__(graph_node)` returns `self.matcher.match_percentage(graph_node.value) >= self.min_match_score` for continuous nodes and `False` otherwise.
    2. `is_equivalent()` returns true if the other node is the same type and has the same the `label`. We do not need to match distributions.
    3. `matches_predicate(other)` returns true if the type matches and the labels match. We do not need to match distributions.
    4. `confirm_match(matched_with)` delegates to the matcher: For continuous feature perception nodes, `self.matcher.update_on_observation(matched_with.value)`, while for continuous feature pattern nodes we do `self.matcher = self.matcher.merge_with(matched_with.matcher)`. Note that for the current plan this means we can't actually merge with other pattern nodes.
8. Add a method `confirm_pattern_match()` to `PerceptionGraphPatternMatch` which mutates `matched_pattern`.
9. Add a method `confirm_graph_match()` to `PerceptionGraphPatternMatch` which mutates `graph_matched_against` when that happens to be a `PerceptionGraphPattern`.
10. Add a keyword argument to `PerceptionGraphPattern.from_graph()` which takes a continuous matching threshold for continuous values.
11. Learners take a continuous matching threshold which they propagate into their nodes. This means:
    1. Add a new attribute `_matching_threshold: float` to `AbstractTemplateLearner`.
    2. Modify each learner to pass this along to `PerceptionGraphPattern.from_graph()`.
12. Learners confirm matches when appropriate. This means:
    1. Modify the subset learner to confirm the match after a successful intersection.
    2. Create and link an issue to handle match confirmation for other learners.

### PR DESCRIPTION
**Pull request by spigo900**
Closes #1119.

This should be mostly in order. I created an issue (#1121) for the one issue left outstanding. I'm planning to blacken the changed files next.

Unit tests are still passing locally AFAICT. No new mypy errors. black is grumpy, wants to reformat *all of the things* <strike>including many files not touched</strike> actually I did change those files when refactoring things, whoops. 🙃

I'd like to test to make sure all this machinery actually works, but I don't think we have any continuous features in the current curriculum sample. I'm planning to construct some very simple examples to test on, which sounds tedious, but I think doable.

### CODE REVIEW

**Comment by spigo900** [comment]
@lichtefeld This should be ready for review. Note I am planning on cleaning up the (very messy) Git history once we're happy with the code.

File: adam/perception/perception_graph.py [hunk 1]
```diff
@@ -1135,6 +1148,33 @@ def intersection(
...
+            graph_logger=graph_logger,
+            ontology=ontology,
+            allowed_matches=allowed_matches,
+            match_mode=match_mode,
+            trim_after_match=trim_after_match,
+        )
+        if attempted_match:
+            return attempted_match.matched_pattern
+        else:
+            return None
```
Comment (lichtefeld) [inline adam/perception/perception_graph.py:1163]:
```suggestion
        return attempted_match.matched_pattern if attempted_match else None
```
Nit: This could be reduced to one line and be just as a clear.

File: adam/perception/perception_graph_nodes.py [hunk 2]
```diff
@@ -81,13 +81,23 @@ class ObjectClusterNode(GraphNode):
...
+
 
 @attrs(frozen=True, slots=True, eq=False)
 class CategoricalNode(GraphNode):
     """A node representing a categorical value feature"""
 
     value: str = attrib(validator=instance_of(str))
 
+    def dot_label(self):
+        return f"CategoricalNode(label=?, value={self.value})"
```
Comment (lichtefeld) [inline adam/perception/perception_graph_nodes.py:99]: Note to self to ensure I update this when I rebase affordances on this PR or vice versa.

Aside: I wish we were using 3.8 so we could just write this as `f"CategoricalNode({label=}, {value=})"` but alas we're not

**Review by lichtefeld — CHANGES_REQUESTED** [review]
Code looks good over all. This is the type of change that makes me wish we could just refactor the entire learner structure again but that's definitely out of scope. 

Have you had any luck with getting travis configured? I'm watching for them to announce that ubuntu 22.04 is available as a base image and would consider updating to that to see if that resolves the issues.

File: adam/experiment/log_experiment.py [hunk 3]
```diff
@@ -348,22 +348,26 @@ def learner_factory_from_params(
                 smoothing_parameter=0.002,
```
[the remark this answers is not in the document]
Comment (spigo900) [reply adam/experiment/log_experiment.py:348]: I thought I did make configurable for the simulated experiments. I may have left it as unconfigurable for the symbolics since it doesn't do anything there.

File: adam/learner/attributes.py [hunk 4]
```diff
@@ -189,6 +190,12 @@ def _update_hypothesis(
             ),
             trim_after_match=pattern_remove_incomplete_region_or_spatial_path,
         )
+        if match:
+            match.confirm_match()
+            return match.intersection
+        # We don't need this, but Mypy wants it.
+        else:
+            return None
```
[the remark this answers is not in the document]
Comment (spigo900) [reply adam/learner/attributes.py:198]: True, will do.

File: adam/experiment/log_experiment.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lichtefeld) [reply adam/experiment/log_experiment.py:348]: Ah you're correct! I missed these were the symbolic matchers 🙃 apologies.

File: adam/perception/perception_graph_nodes.py (hunk 2 above)
Comment (spigo900) [reply adam/perception/perception_graph_nodes.py:99]: Right? I tried to write it that way and PyCharm reminded me we don't have that feature yet. 🙃

File: adam/experiment/log_experiment.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (spigo900) [reply adam/experiment/log_experiment.py:348]: No worries.

File: adam/learner/attributes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (spigo900) [reply adam/learner/attributes.py]: Should be fixed as of 777bb79.

**Comment by spigo900** [comment]
> This is the type of change

Agreed, on both counts. 🙁

> Have you had any luck with getting travis configured?

I haven't, unfortunately. I tried to reproduce the problem locally (I think last week) but so far I was only able to confirm that the setuptools/pip versions in use were not the source of the problem. (Really strange, considering the trace points inside setuptools.)

File: adam/perception/perception_graph_nodes.py (hunk 2 above)
Comment (lichtefeld) [reply adam/perception/perception_graph_nodes.py:99]: I'd love to upgrade to 3.8 but I think something we're using (Vistautils maybe?) was causing problems when I tried. When I updated to 3.7 I tried to just jump to 3.8 but wasn't able to.

**Comment by lichtefeld** [comment]
> I haven't, unfortunately. 

Ok, I'll trust your running of the tests locally for now unless you discover an easy fix. I'm not sure it's worth the extra time investment with the upcoming deliverable.

**Review by lichtefeld — APPROVED** [review]
I'm happy with the code, I trust you'll resolve any local test failures and can cleanup the git history. If you'd like me to run the tests locally as well I'll do so after you've cleaned up the git history just let me know.

**Comment by spigo900** [comment]
@lichtefeld Argh, that reminds me, this code can't run in PyPy because it uses SciPy's implementation of the Gaussian distribution stuff, and PyPy doesn't support SciPy. (What a sentence.) So the new tests can't be run in PyPy. I may have to move them to separate files so we can ignore those files when running with PyPy.

**Comment by spigo900** [comment]
Hmm, actually, it looks like PyPy does support SciPy as of somewhat recently (four years ago). [My information is out of date](https://stackoverflow.com/a/50201186).

**Comment by lichtefeld** [comment]
Ok, we can handle running a few tests in a normal python environment if we end up needing to, hopefully PyPy works though :)

**Comment by spigo900** [comment]
Unfortunately it seems like you can't just pip install scipy for pypy. You have to compile from sources. We'll have to run those tests in CPython.

**Comment by spigo900** [comment]
I've cleaned up the histories and rerun pre-commit things to make sure everything is still good. I also reran the tests locally -- the overall tests with PyPy, and these new tests separately in CPython. Everything seems good. I'm going to merge this.
